### PR TITLE
Add control system trigger

### PR DESCRIPTION
**Pull request by wthrowe**
### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/ControlSystem/Trigger.hpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+namespace control_system {
+/// \ingroup ControlSystemsGroup
+/// \ingroup EventsAndTriggersGroup
+/// Trigger for control system measurements.
+///
+/// This trigger is only intended to be used with the
+/// `control_system::Event` event.  A specialization of this trigger
+/// will be created during control system initialization for each \ref
+/// control_system::protocols::Measurement "measurement".
```
Comment (knelli2) [inline src/ControlSystem/Trigger.hpp]: Maybe add "for each unique measurement" to clarify that even though multiple control systems may use the same measurement, each measurement will only be triggered by one trigger.

File: src/ControlSystem/Trigger.hpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+ public:
+  /// \cond
+  // LCOV_EXCL_START
+  explicit Trigger(CkMigrateMessage* const msg) noexcept : DenseTrigger(msg) {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(Trigger);  // NOLINT
+  // LCOV_EXCL_STOP
+  /// \endcond
+
+  static constexpr bool factory_creatable = false;
```
Comment (knelli2) [inline src/ControlSystem/Trigger.hpp]: Clarifying question: In the documentation above you say that the user has to add the trigger to the "factory creation" struct, but here it's `factory_creatable = false`? Why is that?

File: src/ControlSystem/Trigger.hpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                       const double time) noexcept {
+    return tmpl::as_pack<ControlSystems>([&array_index, &cache, &component,
+                                          &time](
+                                             auto... control_systems) noexcept {
+      return domain::functions_of_time_are_ready<
+          control_system::Tags::MeasurementTimescales>(
+          cache, array_index, component, time,
+          std::array{tmpl::type_from<decltype(control_systems)>::name()...});
+    });
+    return true;
```
Comment (knelli2) [inline src/ControlSystem/Trigger.hpp]: Is this return needed? The above one that returns the `tmpl::as_pack` seems like it should be sufficient.

File: tests/Unit/ControlSystem/Test_EventTriggerMetafunctions.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  static_assert(tmpl::size<result>::value == 2);
+  static_assert(tmpl::list_contains_v<result, expectedB>);
+  static_assert(tmpl::list_contains_v<result, expectedA_option0> or
+                tmpl::list_contains_v<result, expectedA_option1>);
+}
+
+template
+void check<control_system::control_system_events, control_system::Event>();
+template
+void check<control_system::control_system_triggers, control_system::Trigger>();
```
Comment (knelli2) [inline tests/Unit/ControlSystem/Test_EventTriggerMetafunctions.cpp]: GCC doesn't like this. Says `check()` is defined but never used. Maybe rework things so everything is just `static_assert`s or put the function calls inside a `SPECTRE_TEST_CASE`

File: tests/Unit/ControlSystem/Test_Trigger.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+SPECTRE_TEST_CASE("Unit.ControlSystem.Trigger", "[Domain][Unit]") {
+  Parallel::register_classes_with_charm<MeasurementFoT>();
+  using MockRuntimeSystem = ActionTesting::MockRuntimeSystem<Metavariables>;
+  using component = Component<Metavariables>;
+  const component* const component_p = nullptr;
+
+  control_system::Tags::MeasurementTimescales::type measurement_timescales{};
+  measurement_timescales["LabelA"] =
+      std::unique_ptr<MeasurementFoT>{new MeasurementFoT(0.0, {{{3.0}}}, 1.5)};
```
Comment (knelli2) [inline tests/Unit/ControlSystem/Test_Trigger.cpp]: `std::make_unique` here and below

File: tests/Unit/ControlSystem/Test_Trigger.cpp [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+  set_time(0.75);
+
+  CHECK(not trigger.is_ready(box, cache, 0, component_p));
+
+  struct UpdateMeasurementTimescale {
+    static void apply(
+        const gsl::not_null<control_system::Tags::MeasurementTimescales::type*>
+            timescales) noexcept {
+      static_cast<MeasurementFoT&>(*timescales->at("LabelB"))
```
Comment (knelli2) [inline tests/Unit/ControlSystem/Test_Trigger.cpp]: Now that the FunctionOfTime base class has an update function (and the derived class is a PiecewisePolynomial), you don't need to cast to the derived anymore.
`timescales->at("LabelB")->update(0.5, {4.0}, 4.0);` should work just fine.

Also FYI the `UpdateFunctionOfTime` struct in `ControlSystem/UpdateFunctionOfTime.hpp` does this exact mutation.

**Review by knelli2 — CHANGES_REQUESTED** [review]

File: src/ControlSystem/Trigger.hpp (hunk 1 above)
Comment (wthrowe) [reply src/ControlSystem/Trigger.hpp]: I have also updated the text in Event.hpp to match.

File: src/ControlSystem/Trigger.hpp (hunk 2 above)
Comment (wthrowe) [reply src/ControlSystem/Trigger.hpp]: I may have picked a poor name for the factory creation struct and protocol.  In addition to being used by the factory, the lists are used in some places to dispatch to derived classes.  It hadn't occurred to anyone at the time that we would want to dispatch to things that could not be created by a factory.

File: tests/Unit/ControlSystem/Test_Trigger.cpp (hunk 5 above)
Comment (wthrowe) [reply tests/Unit/ControlSystem/Test_Trigger.cpp]: I want to propose changing this guideline eventually, but there's no reason not to continue using `make_unique` in this case, so I'll do it some other time.

**Review by wthrowe — COMMENTED** [review]
Also added a commit fixing a minor bug I noticed while doing the fixups.

File: src/ControlSystem/Trigger.hpp (hunk 2 above)
Comment (knelli2) [reply src/ControlSystem/Trigger.hpp]: Maybe add a comment about this then? I'm sure somebody else in the future will be confused by this.

File: tests/Unit/ControlSystem/Test_Trigger.cpp (hunk 5 above)
Comment (knelli2) [reply tests/Unit/ControlSystem/Test_Trigger.cpp]: I suggested it because clang-tidy was complaining about it

**Review by knelli2 — COMMENTED** [review]
LGTM. You can squash

**Review by knelli2 — DISMISSED** [review]

**Comment by wthrowe** [comment]
Rebased and dropped noexcepts.

**Review by knelli2 — DISMISSED** [review]

**Comment by wthrowe** [comment]
Rebased for merge conflict.

**Review by knelli2 — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**